test/tests/listbox_grouped.js

```diff
@@ -0,0 +1,225 @@
+  await t.context.session.findElement(By.css(`${ex.optionIdBase}2`)).click();
+
+  // Sending the key up arrow will put focus on the first option
+  await listbox.sendKeys(Key.ARROW_UP);
+  await assertAriaSelectedAndActivedescendant(t, ex.listboxSelector, ex.optionSelector, 0);
+});
+
+ariaTest('UP ARROW does not wrap to bottom of listbox', exampleFile, 'key-up-arrow', async (t) => {
+  const listbox = await t.context.session.findElement(By.css(ex.listboxSelector));
+
+  // click the first option
+  await t.context.session.findElement(By.css(`${ex.optionIdBase}1`)).click();
+
+  // arrow up
+  await listbox.sendKeys(Key.ARROW_UP);
+
+  await assertAriaSelectedAndActivedescendant(t, ex.listboxSelector, ex.optionSelector, 0);
+});
+
+ariaTest('UP ARROW moves through groups', exampleFile, 'key-up-arrow', async (t) => {
+  // Sending the key down arrow will put focus on the first option if no options are focused
+  const listbox = await t.context.session.findElement(By.css(ex.listboxSelector));
+  const group1 = await listbox.findElement(By.css('[role="group"]'));
+  const groupOptions = await group1.findElements(By.css('[role="option"]'));
+
+  // click first option in second group
+  await t.context.session.findElement(By.css(`${ex.optionIdBase}${groupOptions.length + 1}`)).click();
+
+  await listbox.sendKeys(Key.ARROW_UP);
+
+  // focus should be on the last option in the first group
+  await assertAriaSelectedAndActivedescendant(t, ex.listboxSelector, ex.optionSelector, groupOptions.length - 1);
+});
+
+ariaTest('HOME moves focus', exampleFile, 'key-home', async (t) => {
+  const listbox = await t.context.session.findElement(By.css(ex.listboxSelector));
+  await listbox.sendKeys(Key.ARROW_DOWN, Key.ARROW_DOWN);
+
+  // Sending key home should always put focus on the first item
+  await listbox.sendKeys(Key.HOME);
+  await assertAriaSelectedAndActivedescendant(t, ex.listboxSelector, ex.optionSelector, 0);
+
+  // Sending key home twice should always put focus on the first item
+  await listbox.sendKeys(Key.HOME);
+  await assertAriaSelectedAndActivedescendant(t, ex.listboxSelector, ex.optionSelector, 0);
+});
+
+ariaTest('END scrolls listbox option into view', exampleFile, 'key-end', async (t) => {
+  const listbox = await t.context.session.findElement(By.css(ex.listboxSelector));
+  const options = await t.context.session.findElements(By.css(ex.optionSelector));
+
+  let listboxBounds = await listbox.getRect();
+  let optionBounds = await options[options.length - 1].getRect();
+
+  t.true(listboxBounds.y + listboxBounds.height - optionBounds.y < 0, 'last option is not initially displayed');
+
+  await listbox.sendKeys(Key.END);
+  listboxBounds = await listbox.getRect();
+  optionBounds = await options[options.length - 1].getRect();
+
+  t.true(listboxBounds.y + listboxBounds.height - optionBounds.y >= 0, 'last option is in view after end key');
+});
+
+ariaTest('Click scrolls listbox option into view', exampleFile, 'key-end', async (t) => {
```

Comment:
I see your desire for a way to tests other attributes of the example :) Good test.

Reply:
Thanks 😊